doc/reference/configuration/configuration_reference.rst

```diff
@@ -2877,6 +2878,120 @@ This section describes configuration parameters related to roles.
 
     **Since:** :doc:`3.0.0 </release/3.0.0>`.
 
+    Specify the roles of an instance.
+    To specify a role's configuration, use the :ref:`roles_cfg <configuration_reference_roles_cfg>` option.
+
+    There are two types of application roles:
+
+    -   Built-in roles. For example, the `CRUD <https://github.com/tarantool/crud>`__ module provides the ``roles.crud-storage`` and ``roles.crud-router`` roles that enable CRUD operations in a sharded cluster.
+    -   Custom roles. For example, you create a custom role to define a stored procedure or implement a supplementary service, such as an e-mail notifier or a replicator.
+
+    To create a custom role, you need to define the following methods and fields:
+
+    -   ``validate()``
+
+        Validate a role's configuration.
+        This method is called when the initial configuration is applied or the configuration is reloaded for the instance with this role.
+        ``validate()`` should throw an error if the validation fails.
+
+    -   ``apply()``
+
+        Apply a role's configuration.
+        This method is called when the initial configuration is applied or the configuration is reloaded for the instance with this role.
+        ``apply()`` should throw an error if the specified configuration can't be applied.
+
+    -   ``stop()``
+
+        Stop a role.
+        This method is called on configuration reload if the role is removed from ``roles`` for the given instance.
+        ``stop()`` should cancel all background fibers and clean up used resources.
+        If these actions can't be performed, ``stop()`` should throw an error.
```

Comment:
I'm not sure that we should push role authors to throw an error from `stop()`.

In my understanding, we should describe the contract, so the author can choose.

Something like the following:

> All these functions (`validate`, `apply` and `stop`) report an unrecoverable error by throwing an error object (usually just a string). An error in any phase of configuration applying means stop of the configuration applying. If role's stop or start raises an error, no roles are stopped or started atferwards. The error is catched and saved. It is shown in alerts reported by the `config:info()` call.